src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java

```diff
@@ -64,70 +102,65 @@ void setProjectId(final String projectId) {
 
     // check that file exists
-    final Path file = artifact == null ? project.getArtifact().getFile().toPath() : artifact;
+    final Path file =
+        artifactPath == null ? project.getArtifact().getFile().toPath() : artifactPath.toPath();
     if (!Files.exists(file)) {
       throw new MojoExecutionException(
           file
               + " does not exist. Make sure to bind the scan goal to a phase that will execute after the artifact to scan has been built");
     }
-    final NewCodeArtifactRequest codeArtifactRequest = NewCodeArtifactRequest.of(projectId, file);
-    getLog().info("Uploading " + file.getFileName() + " to Contrast Scan");
 
-    // create new code artifact
-    final CodeArtifact artifact;
+    final ContrastScanSDK contrastScan = new DefaultContrastScanSDK(contrast, getURL());
+    final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+    final ArtifactScanner scanner =
+        new ArtifactScanner(
+            executor, contrastScan, getOrganizationId(), getProjectId(), POLL_SCAN_INTERVAL);
     try {
-      artifact = contrastScan.createCodeArtifact(getOrganizationId(), codeArtifactRequest);
-    } catch (final UnauthorizedException e) {
-      throw new MojoExecutionException("Failed to authenticate to Contrast", e);
-    } catch (final IOException e) {
-      throw new MojoExecutionException("Failed to upload artifact to Contrast Scan", e);
-    }
+      getLog().info("Uploading " + file.getFileName() + " to Contrast Scan");
+      final ScanOperation operation = scanner.scanArtifact(file, label);
+      getLog().info("Starting scan with label " + label);
 
-    // start scan for new code artifact
-    final StartScanRequest request =
-        StartScanRequest.builder()
-            .projectId(projectId)
-            .codeArtifactId(artifact.getId())
-            .label(label)
-            .build();
-    final Scan scan;
-    try {
-      scan = contrastScan.startScan(getOrganizationId(), request);
-    } catch (final UnauthorizedException e) {
-      throw new MojoExecutionException("Failed to authenticate to Contrast", e);
-    } catch (final IOException e) {
-      throw new MojoExecutionException(
-          "Failed to start scan for code artifact " + artifact.getId(), e);
-    }
+      // show link in build log
+      final URL clickableScanURL = createClickableScanURL(operation.id());
+      getLog().info("Scan results will be available at " + clickableScanURL.toExternalForm());
 
-    // show link in build log
-    final URL clickableScanURL;
-    try {
-      clickableScanURL = createClickableScanURL(scan.getId());
-    } catch (final MalformedURLException e) {
-      throw new MojoExecutionException(
-          "Error building clickable Scan URL. Please contact support@contrastsecurity.com for help",
-          e);
+      // if should not wait, then stop asking for scan results and return
+      if (!waitForResults) {
+        operation.disconnect();
```

Comment:
If createClickableScanURL throws a `MalformedURLException`, we'll never call `operation.disconnect()` here. Seems unlikely to happen, but we should probably consider that case.